src/wireviz/wireviz.py

```diff
@@ -90,108 +58,92 @@ def check_designators(what, where):
-            elif con_con:
-                cocon_coname = list(connections[0].keys())[0]
-                from_pins = expand(connections[0][from_name])
-                to_pins = expand(connections[1][to_name])
-
-                for (from_pin, to_pin) in zip(from_pins, to_pins):
-                    harness.loop(cocon_coname, from_pin, to_pin)
-            if fer_cbl or cbl_fer:
-                from_pins = expand(connections[0][from_name])
-                to_pins = expand(connections[1][to_name])
-
-                if fer_cbl:
-                    ferrule_name = from_name
-                    cable_name = to_name
-                    cable_pins = to_pins
-                else:
-                    ferrule_name = to_name
-                    cable_name = from_name
-                    cable_pins = from_pins
-
-                ferrule_params = yaml_data['ferrules'][ferrule_name]
-                for cable_pin in cable_pins:
-                    ferrule_counter = ferrule_counter + 1
-                    ferrule_id = f'_F{ferrule_counter}'
-                    harness.add_connector(ferrule_id, category='ferrule', **ferrule_params)
-
-                    if fer_cbl:
-                        harness.connect(ferrule_id, 1, cable_name, cable_pin, None, None)
-                    else:
-                        harness.connect(None, None, cable_name, cable_pin, ferrule_id, 1)
 
-        else:
-            raise Exception('Wrong number of connection parameters')
+    def check_designators(what, where): # helper function
+        for i, x in enumerate(what):
+            if x not in yaml_data[where[i]]:
+                return False
+        return True
+
+    autogenerated_ids = {}
+    for connection in yaml_data['connections']:
+        # TODO: check that items are of alternating type CONNECTOR/FERRULE/FERRULE_LIST and CABLE/WIRE
+        # TODO: special case: loops!
+
+        # check that all iterable items (lists and dicts) are the same length
+        itemcount = None
+        for item in connection:
+            if isinstance(item, list):
+                itemcount_new = len(item)
+            elif isinstance(item, dict):
+                if len(item.keys()) != 1:
+                    raise Exception('Dicts may contain only one item here!')
+                itemcount_new = len(expand(list(item.values())[0]))
+            elif isinstance(item, str):
+                continue
+            if itemcount is not None and itemcount_new != itemcount:
+                raise Exception('All lists and dict lists must be the same length!')
+            itemcount = itemcount_new
+        if itemcount is None:
+            raise Exception('No item revealed the number of connections to make!')
+
+        # populate connection list
+        connection_list = []
+        for i, item in enumerate(connection):
+            if isinstance(item, str):  # one single-pin component was specified
```

Comment:
It's the responsibility of the user to make sure it's really a single-pin component. If not, the code doesn't care and simply assumes you want to connect pin 1. Same for the next `elif`.